test/hmr.test.ts

```diff
@@ -0,0 +1,225 @@
+}
+
+const testCases: TestCase[] = [
+  {
+    name: 'simple vue file',
+    path: 'hmr/TestHmr.vue',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content.replace('</script>', '\n</script>')
+    }
+  },
+  {
+    name: 'vue js propagation',
+    path: 'hmr/testHmrPropagation.js',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content.replace('</script>', '\n</script>')
+    }
+  },
+  {
+    name: 'vue dynamic import propagation',
+    path: 'hmr/testHmrPropagationDynamicImport.js',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  },
+  {
+    name: 'vue manual hmr',
+    path: 'hmr/testHmrManual.js',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  },
+  {
+    name: 'vue alias hmr',
+    path: 'alias/aliased/index.js',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  },
+  {
+    name: 'vue css import',
+    path: 'css-@import/testCssAtImportFromScript.css',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  },
+  {
+    name: 'vue css modules import',
+    path: 'css-@import/testCssModuleAtImportFromScript.module.css',
+    expectedMessagesCount: 2,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  },
+  {
+    name: 'vue css style @import hmr',
+    path: 'css-@import/testCssAtImportFromStyle.css',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  },
+  {
+    name: 'vue css style @import modules hmr',
+    path: 'css-@import/testCssModuleAtImportFromStyle.module.css',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  },
+  {
+    name: 'css @import',
+    path: 'css-@import/imported.css',
+    expectedMessagesCount: 1,
+    replacer: (content) => {
+      return content + '\n\n'
+    }
+  }
+]
+
+beforeAll(async () => {
+  try {
+    await fs.remove(tempDir)
+  } catch (e) {}
+  await fs.ensureDir(tempDir)
+  await fs.copy(fixtureDir, tempDir, {
+    filter: (file) => !/dist|node_modules/.test(file)
+  })
+  await execa('yarn', { cwd: tempDir, stdio: 'inherit' })
+  await execa('yarn', {
+    cwd: path.join(tempDir, 'optimize-linked'),
+    stdio: 'inherit'
+  })
+})
+
+afterAll(async () => {
+  try {
+    await fs.remove(tempDir)
+  } catch (e) {}
+})
+
+describe('playground hmr', () => {
+  const baseUrl = `http://localhost:${PORT}`
+
+  const root = tempDir
+
+  let server
+  beforeAll(async () => {
+    server = createServer(config)
+    server.listen(PORT)
+    await once(server, 'listening')
+    // console.log('sleeping')
+    // await sleep(1000000 * 1000)
+  })
+  afterAll(() => {
+    server.close()
+  })
+
+  for (let testCase of testCases) {
+    test(testCase.name, async () => {
+      const traversedFiles = await traverseEsModules({
```

Comment:
```suggestion
      const traversedFiles = await traverseEsModules({
```

Is this redundant?

Reply:
Do you mean the `traversedFiles` variable? i kept the variable in case you need it while debugging the test and know what got traversed, i can remove it if you don't like it